src/resources/adc_esp32.cc

```diff
@@ -104,27 +106,26 @@ PRIMITIVE(init) {
     }
   }
 
-  ADCState* state = _new ADCState(unit, chan);
+  AdcState* state = _new AdcState(group);
   if (!state) MALLOC_FAILED;
 
-  ByteArray* proxy = process->object_heap()->allocate_external_byte_array(0, reinterpret_cast<uint8*>(state), true, false);
+  ByteArray* proxy = process->object_heap()->allocate_proxy();
   if (proxy == null) {
     delete state;
     ALLOCATION_FAILED;
   }
+  state->init(unit, chan);
 
   const int DEFAULT_VREF = 1100;
   esp_adc_cal_characterize(unit, atten, ADC_WIDTH_BIT_12, DEFAULT_VREF, &state->calibration);
 
+  proxy->set_external_address(state);
```

Comment:
This uses template magic to also set the tag on the AdcState so that we can check the type on entry to `get` primitve :-)